openedx/core/djangoapps/course_apps/rest_api/v1/views.py

```diff
@@ -0,0 +1,190 @@
+from edx_api_doc_tools import path_parameter, schema
+from edx_django_utils.plugins import PluginError
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+from edx_rest_framework_extensions.auth.session.authentication import SessionAuthenticationAllowInactiveUser
+from opaque_keys.edx.keys import CourseKey
+from rest_framework import serializers, views
+from rest_framework.exceptions import ValidationError
+from rest_framework.permissions import BasePermission
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from common.djangoapps.student.auth import has_studio_write_access
+from openedx.core.lib.api.authentication import BearerAuthenticationAllowInactiveUser
+from openedx.core.lib.api.view_utils import DeveloperErrorViewMixin, validate_course_key, verify_course_exists
+from ...api import is_course_app_enabled, set_course_app_enabled
+from ...plugins import CourseApp, CourseAppsPluginManager
+
+User = get_user_model()
+log = logging.getLogger(__name__)
+
+
+class HasStudioWriteAccess(BasePermission):
+    """
+    Check if the user has write access to studio.
+    """
+
+    def has_permission(self, request, view):
+        """
+        Check if the user has write access to studio.
+        """
+        user = request.user
+        course_key_string = view.kwargs.get("course_id")
+        course_key = validate_course_key(course_key_string)
+        return has_studio_write_access(user, course_key)
+
+
+class CourseAppSerializer(serializers.Serializer):  # pylint: disable=abstract-method
+    """
+    Serializer for course app data.
+    """
+
+    id = serializers.CharField(read_only=True, help_text="Unique ID for course app.")
+    enabled = serializers.BooleanField(
+        required=True,
+        help_text="Whether the course app is enabled for the specified course.",
+    )
+    name = serializers.CharField(read_only=True, help_text="Friendly name of the course app.")
+    description = serializers.CharField(read_only=True, help_text="A friendly description of what the course app does.")
+    legacy_link = serializers.URLField(required=False, help_text="A link to the course app in the legacy studio view.")
+    allowed_operations = serializers.DictField(
+        read_only=True,
+        help_text="What all operations are supported by the app.",
+    )
+
+    def to_representation(self, instance: CourseApp) -> Dict:
+        course_key = self.context.get("course_key")
+        user = self.context.get("user")
+        data = {
+            "id": instance.app_id,
+            "enabled": is_course_app_enabled(course_key, instance.app_id),
+            "name": instance.name,
+            "description": instance.description,
+            "allowed_operations": instance.get_allowed_operations(course_key, user),
+        }
+        if hasattr(instance, "legacy_link"):
+            data["legacy_link"] = instance.legacy_link(course_key)
```

Comment:
This can be implemented as a concrete method in the base class `CourseApp`. Concrete are not required to be implemented by every child class. So What I am thinking here is to update the `CourseApp` and define a concrete method as folows. 

```
class CourseApp(ABC):
    @classmethod
    def data(cls) -> Dict
        return {}


class ChildClassWithOwnData(CourseApp):
    @classmethod
    def available(cls):
        return True

    @classmethod
    def data(cls): # << name it however you like
        return {"legacy_link": "my-legacy-link"}


class ModernApp(CourseApp):
    def available(self):
        return True

```

The classes that don't define the `data` it'll class parents concrete method. 
The advantage we have here is, if tomorrow if we want to remove this method, we only need to change the respective classes and not everywhere in the code where it is used. So, ultimately, it will reduce the impact of the change.

Reply:
My concern with this is, that we don't have any other use for this 'data' yet, and the scope of this API to explicitly not return any additional app-specific data, and instead leave that to other APIs. So, since there is no scope for future use for this data, what purpose will it serve once legacy_link is removed?

> The advantage we have here is, if tomorrow if we want to remove this method, we only need to change the respective classes and not everywhere in the code where it is used. So, ultimately, it will reduce the impact of the change.

I don't see how this impact how much code change is involved. This is not used everywhere in the code, it isn't designed to be used everywhere in the code. Only the course apps code is aware of this method and uses it. There are only two places where it's used. An if-statement that checks if this method exists, calls it and returns the result, and a serializer field that references it. 

When this is deprecated, you can just remove that if statement, and that's all. The method will stop being used. The classes can still have the method, but it will now be dead code. The main effort is in updating all the classes to remove this method, but that will still need to be done if it's in the data method as described above.

Reply:
Sorry to bring this up again, 
even though it's a small change, but I am not a big fan of objects knowing the internal state functionality of classes. For me, this is violating an Abstraction concept of OOP. 

Yes if tomorrow we decide to remove it, we would change the class not the usage of the class. Today, the usage only exists at one place (herein `CourseAppSerializer`) - but the fact is wherever we need to use these plugins the  `if` condition would be needed there. 

If you are thinking the new class method would have no use we can remove it from the "class" that has defined it thats' it. We wouldn't need to change class usage.


I have another suggestion as well. 
What if we use it in this way, totally _optional_, even if you decide **not** to use the following, I am totally fine with that. 
```
def data(course_key, user)
    return {
            "id": cls.app_id,
            "enabled": is_course_app_enabled(course_key, cls.app_id),
            "name": cls.name,
            "description": cls.description,
            "allowed_operations": cls.get_allowed_operations(course_key, user),
        }
```

@asadazam93 what are your thoughts on this?

Reply:
@awaisdar001 I agree with you. It would be great if that part can be refactored

Reply:
> Sorry to bring this up again, even though it's a small change, but I am not a big fan of objects knowing the internal functionality of classes. For me, this is violating an Abstraction concept of OOP.

I understand what you're saying, but I don't think this is always avoidable. For instance, this is the first iteration of the API. In the next iteration, we might want to add a new property to the class, let's say `app_logo`. If we make this mandatory, then all existing external plugins will break. If we make this optional, then we need an if check. 

This entire plugin concept is not objected oriented to being with. We don't use, `self`, or even `cls` here; so each method on the class could easily be a function disconnected from the class. The class is just acting as a container for these functions. So, the fact that it's a class is just because that's a convenient mechanism to bring this all together. 

If we really want to fully embrace object-oriented design here, we can completely change the design, to take the `course_id` as a class parameter, and then we don't need to pass it to each method. 

> Yes if tomorrow we decide to remove it, we would change the class not the usage of the class. Today, the usage only exists at one place (herein CourseAppSerializer) - but the fact is wherever we need to use these plugins the if condition would be needed there.

> If you are thinking the new class method would have no use tomorrow, we can remove it from the "class" that has defined it thats' it. We wouldn't need to change class usage.

We will explicitly never use these plugins anywhere else.  We are defining a python API in `api.py` and that is where all code that interacts with these plugins will exist. Any other code will need to use these APIs, and should not interact with the plugins directly. Additionally, the only purpose of these APIs is to collect information for the rest APIs so that is the only place the values will be used. 

> What if we use it in this way, totally optional, even if you decide not to use the following, I am totally fine with that.

I am not sure I understand this approach. Do you mean to say all the values should be available via the `data` method? In that case, the class itself has no use. Additionally, there are a few issues:
- We can't use `is_course_app_enabled` in the class since the that itself uses the class to get this value, so it will result in an infinite loop
- Each instance of CourseApp will now need to reimplement this method with the same boilerplate, and instead of simple attribute lookups that fail and raise an error, we will need to do an if check for each of these values.